clients/pkg/promtail/client/batch.go

```diff
@@ -72,19 +72,56 @@ func (b *batch) add(entry api.Entry) error {
 }
 
 func labelsMapToString(ls model.LabelSet, without ...model.LabelName) string {
+	var b strings.Builder
+	var labelsSize int
 	lstrs := make([]string, 0, len(ls))
+
 Outer:
 	for l, v := range ls {
 		for _, w := range without {
 			if l == w {
 				continue Outer
 			}
 		}
-		lstrs = append(lstrs, fmt.Sprintf("%s=%q", l, v))
+
+		str, s := labelNameValueToString(b, l, v)
+		lstrs = append(lstrs, str)
+		labelsSize += s
+	}
+	if len(lstrs) == 0 {
+		return "{}"
 	}
 
 	sort.Strings(lstrs)
-	return fmt.Sprintf("{%s}", strings.Join(lstrs, ", "))
+	return labelListToString(b, labelsSize, lstrs)
+}
+
+// labelListToString replicates 'fmt.Sprintf("{%s}", strings.Join(lstrs, ", "))', but in a more efficent way. We achieve this by using a string.Builder and keeping track of the size at an earlier stage
+func labelListToString(b strings.Builder, labelsSize int, lstrs []string) string {
+	b.Reset()
+	b.Grow(labelsSize + 2 + (2 * (len(lstrs) - 1)))
+	b.WriteString("{")
+	for i := 0; i < len(lstrs)-1; i++ {
+		b.WriteString(lstrs[i])
+		b.WriteString(", ")
+	}
+	// Write the last element seperatly so we have no trailing comma (', ')
+	b.WriteString(lstrs[len(lstrs)-1])
+	b.WriteString("}")
+	return b.String()
+}
+
+// labelNameValueToString replicates 'fmt.Sprintf("%s=%q", l, v)', but in a more efficent way. We achieve this by using a string.Builder. We also return the size for a later concatenation.
+func labelNameValueToString(b strings.Builder, l model.LabelName, v model.LabelValue) (string, int) {
+	// The strings.Builder
+	b.Reset()
+	stringSize := len(l) + len(v) + 3 // 3 for the equals (=) and two quotes (")
+	b.Grow(stringSize)
+	b.WriteString(string(l))
+	b.WriteString("=\"")
+	b.WriteString(string(v))
```

Comment:
:exclamation: Note that this was quoting the value previously, which escaped the quotes within the value: https://go.dev/play/p/1ijtqdBBwJu

You need to do something like:

```go
var (
    quoteBytes := []byte{'"'}
    escapedQuoteBytes := []byte{'"'}
)

func labelNameValueToString(b strings.Builder, l model.LabelName, v model.LabelValue) (string, int) {
    // ...
    b.Write(bytes.Replace(v, quoteBytes, escapedQuoteBytes)
    // ...
}
```

Reply:
Thanks! Good catch. As I realized later and mentioned in [a comment](https://github.com/grafana/loki/pull/7882#discussion_r1045622980) its actually just a key in a map, and it doesn't have to stay the same in this PR.

The place where we turn this map of streams into push requests we discard the key: https://github.com/grafana/loki/blob/05d21106817438f523c6161d0a1312812d6db44c/clients/pkg/promtail/client/batch.go#L125

Reply:
We also [set it in the stream](https://github.com/grafana/loki/blob/b2a85ad7945320765c064836f27b967a2d3fa647/clients/pkg/promtail/client/batch.go#L67-L68) so it's forwarded in the request. (And then that `Labels` field is parsed later)

Reply:
Rather than using `strings.Replace` you might want to check [`strconv.Quote`](https://pkg.go.dev/strconv#Quote) which will properly escape all characters that need to be escaped.

Reply:
I explicitly tried to avoid that as it would cause more allocations here, but you're right in the fact that it quotes more than just a quote rune.

In that case, I'd replace strings.Builder by a byte buffer, and use `strconv.AppendQuote` on that buffer instead, and would do `*(*string)(unsafe.Pointer(&buf))` at the end (which is what `strings.Buidlder` does in its `String()` method.

Reply:
What would be the benefit of that last suggestion? I'm open to adding that, but I do feel like it hurts the readability a bit.

Reply:
It does hurt the readability. The benefit is that we would save yet more allocations (`strconv.Quote` does a few even if there's nothing to change in the quoted string, while `strconv.AppendQuote` can make it in zero allocations)